### PR TITLE
clustermesh-apiserver: extract kvstore client initialization and heartbeat logic in separate cells

### PR DESCRIPTION
**Pull request by giorio94**
This PR extracts the kvstore client initialization and heartbeat logic from the clustermesh-apiserver in two separate cells, to better integrate in the hive paradigm and allow them to be reused by other components. In this phase, the clustermesh-apiserver still relies on a global variable for the kvstore client, which is planned to be removed with subsequent refactoring (in particular, it involves much less changes once https://github.com/cilium/cilium/pull/25049 is in).

```release-note
clustermesh-apiserver: extract kvstore client initialization and heartbeat logic in separate cells
```

### CODE REVIEW

File: pkg/option/config.go [hunk 1]
```diff
@@ -1356,9 +1356,16 @@ func LogRegisteredOptions(vp *viper.Viper, entry *logrus.Entry) {
 	keys := vp.AllKeys()
 	sort.Strings(keys)
 	for _, k := range keys {
-		v := vp.GetStringSlice(k)
-		if len(v) > 0 {
-			entry.Infof("  --%s='%s'", k, strings.Join(v, ","))
+		ss := vp.GetStringSlice(k)
```
Comment (aanm) [inline pkg/option/config.go]: Should we move this commit into a separate PR so that it can be backported to all branches?

**Review by aanm — APPROVED** [review]
just one small nit.

File: pkg/option/config.go (hunk 1 above)
Comment (giorio94) [reply pkg/option/config.go]: It actually seems that `flags.StringToString` should not be used (https://github.com/cilium/cilium/blob/main/contrib/scripts/check-viper.sh#L20-L23). Let me try to figure out how to make `option.NewNamedMapOptions` work for  hive flags. In that way, this commit should no longer be necessary.

**Review by thorn3r — APPROVED** [review]
excellent work! :+1: 
looks like at least one of your CI failures is due to: https://github.com/cilium/cilium/issues/25483

**Review by YutaroHayakawa — APPROVED** [review]
LGTM!

File: pkg/option/config.go (hunk 1 above)
Comment (giorio94) [reply pkg/option/config.go]: I've opened https://github.com/cilium/cilium/pull/25643 which includes the logic to correctly parse StringMap flags through the hive machinery, and the commit to fix the logging of that type of flags. Backporting is not required, since StringMap flags are not currently used (an intermediate wrapper is used to support both json and KV encoding for legacy flags).

**Review by asauber — APPROVED** [review]
LGTM. I appreciate that the temporary global has a good comment.

**Comment by giorio94** [comment]
Rebased onto main to include the changes introduced in https://github.com/cilium/cilium/pull/25643

**Comment by giorio94** [comment]
Rebased onto main to fix a conflict. I've additionally dropped a leftover kvstore initialization, which caused the spurious creation of an additional etcd client:

```diff
--- a/clustermesh-apiserver/main.go
+++ b/clustermesh-apiserver/main.go
@@ -576,21 +588,19 @@ func startServer(startCtx hive.HookContext, clientset k8sClient.Clientset, servi
 	}
 
 	var err error
-	if err = kvstore.Setup(context.Background(), "etcd", option.Config.KVStoreOpt, nil); err != nil {
-		log.WithError(err).Fatal("Unable to connect to etcd")
-	}
 
 	config := cmtypes.CiliumClusterConfig{
 		ID: cfg.clusterID,
 	}
@@ -602,6 +612,7 @@ func startServer(startCtx hive.HookContext, clientset k8sClient.Clientset, servi
 	}
 
 	ciliumNodeStore, err = store.JoinSharedStore(store.Configuration{
+		Backend:    backend,
 		Prefix:     nodeStore.NodeStorePrefix,
 		KeyCreator: nodeStore.KeyCreator,
 	})
```
